<a href="https://colab.research.google.com/github/jawariyakhawer/knn-fraud./blob/main/Copy_of_knn_fraud.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

In [ ]:
df = pd.read_csv("creditcard.csv")

df.head()

,Time,V1,V2,V3,V4,V5,V6,V7,V8,V9,...,V21,V22,V23,V24,V25,V26,V27,V28,Amount,Class
0,0,-1.359807,-0.072781,2.536347,1.378155,-0.338321,0.462388,0.239599,0.098698,0.363787,...,-0.018307,0.277838,-0.110474,0.066928,0.128539,-0.189115,0.133558,-0.021053,149.62,0.0
1,0,1.191857,0.266151,0.166480,0.448154,0.060018,-0.082361,-0.078803,0.085102,-0.255425,...,-0.225775,-0.638672,0.101288,-0.339846,0.167170,0.125895,-0.008983,0.014724,2.69,0.0
2,1,-1.358354,-1.340163,1.773209,0.379780,-0.503198,1.800499,0.791461,0.247676,-1.514654,...,0.247998,0.771679,0.909412,-0.689281,-0.327642,-0.139097,-0.055353,-0.059752,378.66,0.0
3,1,-0.966272,-0.185226,1.792993,-0.863291,-0.010309,1.247203,0.237609,0.377436,-1.387024,...,-0.108300,0.005274,-0.190321,-1.175575,0.647376,-0.221929,0.062723,0.061458,123.50,0.0
4,2,-1.158233,0.877737,1.548718,0.403034,-0.407193,0.095921,0.592941,-0.270533,0.817739,...,-0.009431,0.798278,-0.137458,0.141267,-0.206010,0.502292,0.219422,0.215153,69.99,0.0


In [ ]:
print("Shape:", df.shape)
print()

df.info()

Shape: (53571, 31)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 53571 entries, 0 to 53570
Data columns (total 31 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Time    53571 non-null  int64  
 1   V1      53571 non-null  float64
 2   V2      53571 non-null  float64
 3   V3      53571 non-null  float64
 4   V4      53571 non-null  float64
 5   V5      53571 non-null  float64
 6   V6      53571 non-null  float64
 7   V7      53571 non-null  float64
 8   V8      53571 non-null  float64
 9   V9      53571 non-null  float64
 10  V10     53571 non-null  float64
 11  V11     53571 non-null  float64
 12  V12     53571 non-null  float64
 13  V13     53571 non-null  float64
 14  V14     53571 non-null  float64
 15  V15     53571 non-null  float64
 16  V16     53571 non-null  float64
 17  V17     53571 non-null  float64
 18  V18     53571 non-null  float64
 19  V19     53571 non-null  float64
 20  V20     53571 non-null  float64
 21  V21     53571 n

In [ ]:
df.describe()

,Time,V1,V2,V3,V4,V5,V6,V7,V8,V9,...,V21,V22,V23,V24,V25,V26,V27,V28,Amount,Class
count,53571.000000,53571.000000,53571.000000,53571.000000,53571.000000,53571.000000,53571.000000,53571.000000,53571.000000,53571.000000,...,53571.000000,53571.000000,53570.000000,53570.000000,53570.000000,53570.000000,53570.000000,53570.000000,53570.000000,53570.000000
mean,30004.864404,-0.243182,0.000586,0.691588,0.179664,-0.257872,0.102917,-0.119311,0.052872,0.094530,...,-0.028932,-0.106714,-0.039090,0.007830,0.135096,0.020158,0.003957,0.003972,94.260870,0.002856
std,13302.545313,1.879598,1.639428,1.483708,1.393856,1.411097,1.313136,1.271570,1.214929,1.201748,...,0.731807,0.636939,0.593273,0.596080,0.438590,0.500092,0.388427,0.329538,252.290845,0.053366
min,0.000000,-56.407510,-72.715728,-32.965346,-5.172595,-42.147898,-26.160506,-26.548144,-41.484823,-9.283925,...,-20.262054,-8.593642,-26.751119,-2.836627,-7.495741,-1.577118,-8.567638,-9.617915,0.000000,0.000000
25%,23663.500000,-0.994263,-0.573628,0.212411,-0.724005,-0.873897,-0.636806,-0.606014,-0.147112,-0.630279,...,-0.230582,-0.528974,-0.179410,-0.322721,-0.128328,-0.330573,-0.063648,-0.006589,7.680000,0.000000
50%,34251.000000,-0.249472,0.078153,0.790661,0.189575,-0.290113,-0.152437,-0.075304,0.058859,-0.012999,...,-0.066124,-0.082370,-0.051736,0.062416,0.174351,-0.073699,0.008850,0.022405,25.495000,0.000000
75%,40127.000000,1.154647,0.732455,1.424778,1.061933,0.281788,0.493733,0.425496,0.333521,0.780864,...,0.110098,0.307953,0.079102,0.401967,0.421204,0.298577,0.083071,0.076498,86.867500,0.000000
max,46050.000000,1.960497,18.183626,4.101716,16.491217,34.801666,22.529298,36.677268,20.007208,10.392889,...,22.614889,5.805795,17.297845,4.014444,5.525093,3.517346,11.135740,33.847808,12910.930000,1.000000


In [ ]:
df.isnull().sum()

,0
Time,0
V1,0
V2,0
V3,0
V4,0
V5,0
V6,0
V7,0
V8,0
V9,0


In [ ]:
# Check missing values again
print(df.isnull().sum())

# Fill missing values with column mean
df = df.fillna(df.mean(numeric_only=True))

# Verify
print("\nAfter Filling Missing Values:\n")
print(df.isnull().sum())

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       1
V24       1
V25       1
V26       1
V27       1
V28       1
Amount    1
Class     1
dtype: int64

After Filling Missing Values:

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64


In [ ]:
df.fillna(df.mean(numeric_only=True), inplace=True)

In [ ]:
# Features and Target
X = df.drop("Class", axis=1)
y = df["Class"].astype(int) # Convert target to integer type

In [ ]:
# Feature Scaling
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X = scaler.fit_transform(X)

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(X_train, y_train)

KNeighborsClassifier()

In [ ]:
# Prediction
y_pred = knn.predict(X_test)

In [ ]:
# Accuracy
from sklearn.metrics import accuracy_score

print("Accuracy:", accuracy_score(y_test, y_pred))

Accuracy: 0.9975734951003267


In [ ]:
# Confusion Matrix & Report
from sklearn.metrics import confusion_matrix, classification_report

print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))

[[10689     0]
 [   26     0]]
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     10689
           1       0.00      0.00      0.00        26

    accuracy                           1.00     10715
   macro avg       0.50      0.50      0.50     10715
weighted avg       1.00      1.00      1.00     10715



/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


# Conclusion

The K-Nearest Neighbors (KNN) algorithm was successfully implemented for credit card fraud detection.

The dataset was preprocessed by handling missing values and scaling the features. The model achieved an accuracy of 99.76%, indicating excellent performance in classifying transactions.

The confusion matrix and classification report show that the model effectively distinguishes between normal and fraudulent transactions, making KNN a suitable algorithm for this classification task.